common/file/archive/zip.go

```diff
@@ -0,0 +1,178 @@
+package archive
+
+import (
+	"archive/zip"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	log "github.com/thrasher-corp/gocryptotrader/logger"
+)
+
+const (
+	// ErrUnableToCloseFile message to display when file handler is unable to be closed normally
+	ErrUnableToCloseFile string = "Unable to close file %v %v"
+)
+
+var (
+	addFilesToZip func(z *zip.Writer, src string, isDir bool) error
+)
+
+func init() {
+	addFilesToZip = addFilesToZipWrapper
+}
+
+// UnZip extracts input zip into dest path
+func UnZip(src, dest string) (fileList []string, err error) {
+	z, err := zip.OpenReader(src)
+	if err != nil {
+		return
+	}
+
+	for x := range z.File {
+		fPath := filepath.Join(dest, z.File[x].Name) // nolint:gosec
+		// We ignore gosec linter above because the code below files the file traversal bug when extracting archives
+		if !strings.HasPrefix(fPath, filepath.Clean(dest)+string(os.PathSeparator)) {
+			err = z.Close()
+			if err != nil {
+				log.Errorf(log.Global, ErrUnableToCloseFile, z, err)
+			}
+			err = fmt.Errorf("%s: illegal file path", fPath)
+			return
+		}
+
+		if z.File[x].FileInfo().IsDir() {
+			err = os.MkdirAll(fPath, os.ModePerm)
+			if err != nil {
+				return
+			}
+			continue
+		}
+
+		err = os.MkdirAll(filepath.Dir(fPath), 0770)
+		if err != nil {
+			return
+		}
+
+		var outFile *os.File
+		outFile, err = os.OpenFile(fPath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, z.File[x].Mode())
+		if err != nil {
+			return
+		}
+
+		var eFile io.ReadCloser
+		eFile, err = z.File[x].Open()
+		if err != nil {
+			err = outFile.Close()
+			if err != nil {
+				log.Errorf(log.Global, ErrUnableToCloseFile, outFile, err)
+			}
+			return
+		}
+
+		_, errIOCopy := io.Copy(outFile, eFile)
+		if errIOCopy != nil {
+			err = z.Close()
+			if err != nil {
+				log.Errorf(log.Global, ErrUnableToCloseFile, z, err)
+			}
+			err = outFile.Close()
+			if err != nil {
+				log.Errorf(log.Global, ErrUnableToCloseFile, outFile, err)
+			}
+			err = eFile.Close()
+			if err != nil {
+				log.Errorf(log.Global, ErrUnableToCloseFile, eFile, err)
+			}
+			return fileList, errIOCopy
+		}
+		err = outFile.Close()
+		if err != nil {
+			log.Errorf(log.Global, ErrUnableToCloseFile, outFile, err)
+		}
+		err = eFile.Close()
+		if err != nil {
+			log.Errorf(log.Global, ErrUnableToCloseFile, eFile, err)
+		}
+		if err != nil {
+			return
+		}
+
+		fileList = append(fileList, fPath)
+	}
+	return fileList, z.Close()
+}
+
+// Zip archives requested file or folder
+func Zip(src, dest string) error {
+	i, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Create(dest)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	z := zip.NewWriter(f)
+	defer z.Close()
+
+	var dir bool
+	if i.IsDir() {
+		dir = true
+	}
+	err = addFilesToZip(z, src, dir)
+	if err != nil {
+		z.Close()
```

Comment:
I think this might be handled in the [defer](https://github.com/thrasher-corp/gocryptotrader/pull/411/files#diff-5bd2c85b54602c554404dd7cd374dae5R122)

Reply:
I close z here because i attempt to remove the broken zip after a failure and want to make sure there is no open FD  (HELLO WINDOWS USERS)